### ISSUE TITLE
Tutorial: How to use 'newFilter' and 'filterChanges'?

### ISSUE THREAD

**Issue by kabaluyot** [ethereum/go-ethereum#27315]
How to use the 'newFilter' and 'filterChanges' for listening to contract event logs?

**Comment by s1na** [comment]
I recommend using the log subscription instead if your client supports websocket. Here's an example how you'd do it with `ethers.js`:

```js
const ethers = require('ethers')

async function main() {
    const provider = new ethers.providers.WebSocketProvider()
    provider.on({ address: '0xabc', topics: [] }, (log) => {
        console.log(txhash)
    })
}

main().then(console.log).catch(console.error)
```

**Comment by kabaluyot** [comment]
Yes -- was doing this all the time using `SubscribeFilterLogs` with websocket.

But other clients also use that `newFilter` and `filterChanges` as it is done via long polling and forum's says it is ideal for critical indexers ..

**Comment by s1na** [comment]
Oh I thought all clients support the subscription? Do you have a link to the forum post?

Anyhow, this is a sample how you can use it:

```js
const ethers = require("ethers");
const filter = {
  fromBlock: "0xe20360",
  toBlock: "0xe20411",
  address: "0x6B175474E89094C44Da98b954EedeAC495271d0F",
  topics: [],
};
(async () => {
  const provider = new ethers.providers.JsonRpcProvider();
  const filterId = await provider.send("eth_newFilter", [filter]);
  console.log(filterId);
})();
```

This will register a filter. Then you will have to poll `eth_getFilterChanges` to get result back regularly:

```js
const ethers = require("ethers");
(async () => {
  const provider = new ethers.providers.JsonRpcProvider();
  const logs = await provider.send("eth_getFilterChanges", ["YOUR_FILTER_ID"]);
  console.log(logs);
})();
```

**Comment by kabaluyot** [comment]
Basically I want to use that eth_newFilter and eth_getFilterChanges using this go-ethereum client but can't find any samples

**Comment by kabaluyot** [comment]
What I mean is using this geth client itself to call eth_newFilter and eth_getFilterChanges.

My use case is I have a Go API that listens to contract events from an external RPC. Aside from using `SubscribeFilters`, I want to use that filters.

**Comment by kabaluyot** [comment]
Something like this as a client: https://goethereumbook.org/en/event-subscribe/

**Comment by kabaluyot** [comment]
From my opinion, it should be supported since most clients do support it too. It's forcing me atm to implement it manually or look for other ethclient just to get the support.

Also long polling with newFilter + getFilterChanges combo is a good alternative or a backup for subscription for ensuring no events are missed. From my experience when using most RPC providers (free or dedicated nodes), there are really chances when websocket abnormally closes at the middle of transactions resulting to missed events.

**Comment by fjl** [comment]
The ethclient API does not directly wrap the provided RPC APIs. As we see it, subscriptions are superior way to get this information.  

However, it would be quite useful to add a fallback subscription based on getFilterChanges in ethclient. This subscription implementation would be used when the server does not support subscriptions. It would basically poll the filter and deliver the results into the subscription channel.

**Comment by fjl** [comment]
Trying out an implementation of the fallback subscription here: https://github.com/ethereum/go-ethereum/pull/27541

**Comment by kabaluyot** [comment]
Yes the fallback to filter is nice to have. 

But is it really not possible to have this wrappers available https://github.com/ethereum/go-ethereum/pull/27541 in ethclient (even still deliver result to subscription)? 

For example in my use case I got this subscription via websocket listener but I want to open another listener using the filter polling which have delay say 30s .. Reason is in the event of abnormal websocket closure (sure there is some reconnection happening here but some events might be missed), the filter polling listener will pick up missed transactions

**Comment by MariusVanDerWijden** [comment]
I think this feature is not really necessary anymore. All clients support subscriptions now afaik which as Felix said is the superior solution. I think this can be closed
